File: 2022/CSharp/day01/Program.cs
namespace Day01
{

    struct Elf
    {
        public List<int> Calories { get; private set; }

        public Elf()
        {
            Calories = new List<int>();
        }
    }

    class Program
    {

        static void Main(string[] args)
        {
            string input = File.ReadAllText("./input.txt");
            Part1(input);
            Part2(input);
        }

        static List<Elf> ProcessInput(string input)
        {
            string[] elfStrings = input.Split("\n\n");
            List<Elf> elves = new List<Elf>();

            foreach (string elfString in elfStrings)
            {
                Elf elf = new Elf();
                string[] calorieStrings = elfString.Split("\n");
                foreach (string calorieString in calorieStrings)
                {
                    if (calorieString.Length == 0)
                    {
                        continue;
                    }
                    elf.Calories.Add(int.Parse(calorieString.Trim()));
                }
                elves.Add(elf);
            }

            return elves;
        }

        static void Part1(string input)
        {
            var elves = ProcessInput(input);
            Elf maxElf = (from elf in elves
                          orderby elf.Calories.Sum() descending
                          select elf).First();
            Console.WriteLine($"Part 1: {maxElf.Calories.Sum()}");
        }

        static void Part2(string input)
        {
            var elves = ProcessInput(input);
            var elfMax = (from elf in elves
                          orderby elf.Calories.Sum() descending
                          select elf).Take(3).Sum(elf => elf.Calories.Sum());
            Console.WriteLine($"Part 2: {elfMax}");
        }

    }
}

File: 2022/FSharp/day01/Program.fs
let readLines filePath = System.IO.File.ReadLines(filePath)

let splitInput lines: string list list =
    match lines with
    | [] -> []
    | lines ->
        let rec splitInputInternal remainingLines acc =
            match remainingLines with
            | [] -> [ acc ]
            | "" :: rest -> acc :: (splitInputInternal rest [])
            | x :: rest -> splitInputInternal rest (x :: acc)
        
        splitInputInternal lines []


let part1 =
    "input.txt"
    |> readLines
    |> Seq.toList
    |> splitInput
    |> List.map (List.map int)
    |> List.map (List.fold (+) 0)
    |> List.max

let part2 =
    "input.txt"
    |> readLines
    |> Seq.toList
    |> splitInput
    |> List.map (List.map int)
    |> List.map (List.fold (+) 0)
    |> List.sortDescending
    |> Seq.take 3
    |> Seq.sum

printfn "Part 1: %d" part1 
printfn "Part 2: %d" part2 